expression/builtin_arithmetic.go

```diff
@@ -548,6 +552,142 @@ func (s *builtinArithmeticDivideDecimalSig) evalDecimal(row []types.Datum) (*typ
 	return c, false, err
 }
 
+type arithmeticModFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *arithmeticModFunctionClass) setType4ModRealOrDecimal(retTp, a, b *types.FieldType, isDecimal bool) {
+	retTp.Decimal = int(math.Max(float64(a.Decimal), float64(b.Decimal)))
+	if isDecimal && retTp.Decimal > mysql.MaxDecimalScale {
+		retTp.Decimal = mysql.MaxDecimalScale
+	}
+
+	retTp.Flen = int(math.Max(float64(a.Flen), float64(b.Flen)))
+
+	if isDecimal {
+		retTp.Flen = int(math.Min(float64(retTp.Flen), float64(mysql.MaxDecimalWidth)))
+		return
+	}
+	retTp.Flen = int(math.Min(float64(retTp.Flen), float64(mysql.MaxRealWidth)))
```

Comment:
if both `a.Flen` and `b.Flen` equal to  types.UnspecifiedLength, which is 0, `retTp.Flen` is -1 ?

Reply:
So when both `a.Flen` and `b.Flen` equal to types.UnspecifiedLength, `retTp.Flen` should return 0?

Reply:
no, we can let `retTp.Flen` to be `types.UnspecifiedLength` in this case.

Reply:
That's the same with me. This part code did the same effect in such scenario.
At first, `retTp.Flen` save the maximum  of `a.Flen` and `b.Flen`. When both  `a.Flen` and `b.Flen` equal to types.UnspecifiedLength, `retTp.Flen`'s value is types.UnspecifiedLength too. Then to avoid `retTp.Flen` exceed the limit of the type's maximum lenth, adjust its value. No matter the `mysql.MaxRealWidth` or `mysql.MaxDecimalWidth`,  `retTp.Flen`(which is types.UnspecifiedLength now) is the minimum of them. So the result is types.UnspecifiedLength.

Reply:
Sorry, it's my fault. Actually,  what I'm trying to state is:
- when one of `a.Decimal` and `b.Decimal` is `types.UnspecifiedLength`, `retTp.Decimal` should be `types.UnspecifiedLength`, and `retTp.Flen` should be `types.UnspecifiedLength` as well
- when one of `a.Flen` and `b.Flen` is `types.UnspecifiedLength`，`retTp.Flen` should be `types.UnspecifiedLength`

Reply:
gotcha